src/virtual_packages.cpp

```diff
@@ -0,0 +1,201 @@
+            // Note: we could also inspect /System/Library/CoreServices/SystemVersion.plist which is
+            // an XML file
+            //       that contains the same information. However, then we'd either need an xml
+            //       parser or some other crude method to read the data
+            std::vector<std::string> args = { "sw_vers", "-productVersion" };
+            auto [status, ec] = reproc::run(
+                args, reproc::options{}, reproc::sink::string(out), reproc::sink::string(err));
+
+            if (ec)
+            {
+                LOG_ERROR
+                    << "Could not find macOS version by calling 'sw_vers -productVersion'\nPlease file a bug report.\nError: "
+                    << ec.message();
+                return "";
+            }
+            return std::string(strip(out));
+        }
+
+        std::string glibc_version()
+        {
+            const char* version = "";
+#ifndef _WIN32
+            // TODO figure out the paths for this
+            void* handle = dlopen("libc.so.6", RTLD_LAZY);
+
+            if (!handle)
+            {
+                LOG_DEBUG << "Cannot find libc library: " << dlerror() << '\n';
+                return "";
+            }
+
+            // reset errors
+            dlerror();
+
+            typedef const char* (*func_t)(void);
+            auto* gnu_get_libc_version = (func_t)(dlsym(handle, "gnu_get_libc_version"));
+
+            const char* dlsym_error = dlerror();
+
+            if (dlsym_error)
+            {
+                LOG_DEBUG << "Cannot load symbol 'gnu_get_libc_version': " << dlsym_error;
+                dlclose(handle);
+                return "";
+            }
+
+            version = gnu_get_libc_version();
+            dlclose(handle);
+#endif
+            return std::string(strip(version));
+        }
+
+        std::string cuda_version()
+        {
+#ifdef _WIN32
+            std::string lib_filename = "nvcuda.dll";
+            // TODO: implement for Windows
+
+#else
+            std::vector<std::string> lib_filenames;
+
+            if (on_linux)
+            {
+                lib_filenames = {
+                    "libcuda.so",                           // check library path first
+                    "/usr/lib64/nvidia/libcuda.so",         // Redhat/CentOS/Fedora
+                    "/usr/lib/x86_64-linux-gnu/libcuda.so"  // Ubuntu
+                };
+            }
+            else if (on_mac)
+            {
+                lib_filenames = { "libcuda.dylib",  // check library path first
+                                  "/usr/local/cuda/lib/libcuda.dylib" };
+            }
+
+            for (auto lib : lib_filenames)
+            {
+                void* handle = dlopen(lib.c_str(), RTLD_NOW);
+                if (handle)
+                {
+                    LOG_DEBUG << "CUDA DLL found at " << lib;
+                    dlerror();
+
+                    int ret;
+
+                    int (*cuInit)(u_int);
```

Comment:
Thanks for the pointer @obilaniu , it should be fixed now.
Let me know if you still get errors on Alpine.

Reply:
@adriendelsalle Yes, the type issue is now resolved.

A personal interest of mine is having fully-static `micromamba` builds, and to do that I build on Alpine Linux with musl-libc, and with as much as possible LTOed. My builds are currently smaller than the dynamic Glibc ones, as a point of comparison (#572). My desire for a fully-static link stems from a desire to (at least theoretically) be able to bootstrap an entire system from a `micromamba` call; It's as powerful as any package manager and even uses the same `libsolv` as openSUSE.

This PR poses some difficulties for me because `dlopen()` is essentially incompatible with static linking. However, I'm unable to formulate an easy way to bypass that; My current ideas revolve around executing `/lib64/ld-linux-x86-64.so.2 libc.so.6` or the architectural equivalent, which seemingly everywhere appears to print the version of the Glibc C library in the first line. For the CUDA detection, I have no real idea outside `nvidia-smi -q | grep 'CUDA Version'`, but that's ugly. I don't want to obstruct this project and its efforts at achieving compatibility with Conda though, so I will have to think hard about whether it's even possible to maintain such an absolute level of system isolation.

Reply:
@obilaniu thanks for actively monitoring this here, and thinking with us!

I didn't fully think through the implications of loading dynamic libraries like this here so it's very welcome that you're pointing out that fully static builds will be problematic. I agree that calling / attempting to call an external function to find the version numbers might be easier, and I would be quite happy to go down that route but as you said, the nvidia-smi interface is not really nice and for glibc the printed string seemed to look quite different between distros.

- We can always make this a compile-time optional feature (WITH_DYNAMIC_PACKAGE_DETECTION or something)
- We can (later) ship another package that only finds virtual packages (maintained as micromamba-virtual-packages-finder), installable through micromamba, obviously (seperate binary)
- We can define virtual packages using YAML or JSON and put them into some RC file or other config directory (this would enable compatibility even without properly detecting stuff)

Again, thanks for coming to this PR.

Reply:
I had a look at the confstr function to detect the libc version, but unfortunately musl seems to only implement a stub:

```
>>> import os
>>> os.confstr(2)
'glibc 2.32'
```

and on Alpine (using native Python):

```
>>> os.confstr(0)
'/bin:/usr/bin'
>>> os.confstr(1)
''
>>> os.confstr(2)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: [Errno 22] Invalid argument
```

It's possible to "execute" libc:

```
/ # ./lib/libc.musl-x86_64.so.1 
musl libc (x86_64)
Version 1.1.24
Dynamic Program Loader
Usage: ./lib/libc.musl-x86_64.so.1 [options] [--] pathname [args]
```

and with glibc:

```
(base) ~/Programs:$ /lib64/libc.so.6 
GNU C Library (GNU libc) stable release version 2.32.
Copyright (C) 2020 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.
Compiled by GNU CC version 10.2.1 20201125 (Red Hat 10.2.1-9).
libc ABIs: UNIQUE IFUNC ABSOLUTE
For bug reporting instructions, please see:
<https://www.gnu.org/software/libc/bugs.html>.
```

But obviously this is going to be a nightmare to parse.

Reply:
One last point: I think for conda it was already discussed to make virtual packages plugin based. i think that is the ideal scenario, although a bit tricky without python.

Reply:
@wolfv It's less bad than it first appears because essentially every arch-dependent `conda` package that ships a dynamically-linked binary bakes into itself the assumption that the Glibc dynamic linker exists and is at a certain path - because the `conda` compilers make that assumption too. `readelf -l` over your own dynamically-linked `micromamba` builds reveals that the `PT_INTERP` tags are:

- `/lib/ld-linux-aarch64.so.1` (Aarch64)
- `/lib64/ld64.so.2` (PPC64LE)
- `/lib64/ld-linux-x86-64.so.2` (x86-64)

Only one of these strings would need to be baked into `micromamba` - the one for the arch being built. These values are going to change only once in a blue moon, or never.

As it so happens, Glibc `libc.so.6` is designed, on all Linux systems, to be _directly executable_, printing a version. The magic of `/lib64/ld-linux-x86-64.so.2 libc.so.6` is that it offloads to Glibc both the job of finding `libc.so.6` in the usual location following exactly the default Glibc rules, _and_ by a happy coincidence attempts to execute it once it has done so, which prints version information.

Near as I can tell, the Glibc version is always in the first line, after the word `version`. Example printouts from systems I have access to:

- `GNU C Library (GNU libc) stable release version 2.26 (git 1c9a5c270d8b), by Roland McGrath et al.`
- `GNU C Library (Ubuntu GLIBC 2.27-3ubuntu1.2) stable release version 2.27.`
- `GNU C Library (GNU libc) stable release version 2.17, by Roland McGrath et al.`
- `GNU C Library (GNU libc) stable release version 2.24, by Roland McGrath et al.`

In each and every x86-64 system I have visited, it also appears that `/lib64/ld-linux-x86-64.so.2` is a symlink with a relative (`ld-2.27.so`) or absolute (`/lib/x86_64-linux-gnu/ld-2.27.so`) path. Parsing the `readlink()` could therefore be viable.

Since CUDA and `libcuda.so` is always built against Glibc, the `__cuda` package's existence/usefulness is predicated on `__glibc`. If `__glibc` is detected, then obviously a certain minimum Glibc-supporting system pre-exists, and at this point some Glibc tool can be downloaded, or else `micromamba` can resort to some other strategy. Or it could receive that information as a special `--arg`.